display_list/display_list_builder.h

```diff
@@ -376,65 +381,131 @@ class DisplayListBuilder final : public virtual Dispatcher,
     // This offset is only valid if |has_layer| is true.
-    size_t save_layer_offset;
-
-    bool has_deferred_save_op_ = false;
+    size_t save_layer_offset() const { return save_layer_offset_; }
 
-    bool has_layer;
-    bool cannot_inherit_opacity;
-    bool has_compatible_op;
+    bool has_layer() const { return has_layer_; }
+    bool cannot_inherit_opacity() const { return cannot_inherit_opacity_; }
+    bool has_compatible_op() const { return cannot_inherit_opacity_; }
+    SkM44& matrix() { return matrix_; }
+    SkMatrix& matrix33() { return matrix33_; }
+    SkRect& clip_bounds() { return clip_bounds_; }
 
-    SkM44 matrix;
-    SkRect clip_bounds;
+    void update_matrix33() { matrix33_ = matrix_.asM33(); }
 
-    bool is_group_opacity_compatible() const { return !cannot_inherit_opacity; }
+    bool is_group_opacity_compatible() const {
+      return !cannot_inherit_opacity_;
+    }
 
-    void mark_incompatible() { cannot_inherit_opacity = true; }
+    void mark_incompatible() { cannot_inherit_opacity_ = true; }
 
     // For now this only allows a single compatible op to mark the
     // layer as being compatible with group opacity. If we start
     // computing bounds of ops in the Builder methods then we
     // can upgrade this to checking for overlapping ops.
     // See https://github.com/flutter/flutter/issues/93899
     void add_compatible_op() {
-      if (!cannot_inherit_opacity) {
-        if (has_compatible_op) {
-          cannot_inherit_opacity = true;
+      if (!cannot_inherit_opacity_) {
+        if (has_compatible_op_) {
+          cannot_inherit_opacity_ = true;
         } else {
-          has_compatible_op = true;
+          has_compatible_op_ = true;
         }
       }
     }
+
+    // The filter to apply to the layer bounds when it is restored
+    std::shared_ptr<const DlImageFilter> filter() { return filter_; }
+
+    // is_unbounded should be set to true if we ever encounter an operation
+    // on a layer that either is unrestricted (|drawColor| or |drawPaint|)
+    // or cannot compute its bounds (some effects and filters) and there
+    // was no outstanding clip op at the time.
+    // When the layer is restored, the outer layer may then process this
+    // unbounded state by accumulating its own clip or transferring the
+    // unbounded state to its own outer layer.
+    // Typically the DisplayList will have been constructed with a cull
+    // rect which will act as a default clip for the outermost layer and
+    // the unbounded state of all sub layers will eventually be caught by
+    // that cull rect so that the overall unbounded state of the entire
+    // DisplayList will never be true.
+    //
+    // SkPicture treats these same conditions as a Nop (they accumulate
+    // the SkPicture cull rect, but if it was not specified then it is an
+    // empty Rect and so has no effect on the bounds).
+    // If the Calculator object accumulates this flag into the root layer,
```

Comment:
Are these references to Calculator still valid (this entire paragraph). Has this functionality moved here from the existing Calculator helper?

Reply:
These comments are moved from where they were before. I don't quite know how to tweak it, it's a bit difficult for me, would you mind helping to tweak it? thanks!

Reply:
Remove it now.